### DISCUSSION TITLE
Collision avoidance script based flying

### DISCUSSION THREAD

**Discussion by heinrichwagner** [bitcraze/discussions#890] [Q&A] [answered]
Hi all, 

I would like to use the collision avoidance function out of the decentralized swarm demo from the Branch “ICRA2023” (crazyflie-firmware-experimental). 
 
The main idea is to have this function always on, while flying with a python script and cflib from a machine through the crazyradio dongle. 
 
I moved the enableCollisionAvoidance() method to the state “wait for take off” in pilot.c :
 
```
case STATE_WAIT_FOR_TAKE_OFF:        // This is the main state when not flying

            enableCollisionAvoidance();   // moved from state "taking off"

            if (! chargedForTakeoff()){
                //do nothing, wait for the battery to be charged
            }
            else if (needMoreCopters(state)){
                random_time_for_next_event_ms = get_next_random_timeout(now_ms);
                DEBUG_PRINT("More copters needed, preparing for take off...\n");
                state = STATE_PREPARING_FOR_TAKE_OFF;
            }
            break;
```
 
Our procedure to test the collision avoidance with two drones and the above mentioned change in the firmware:
 
Wait for the drones to reach the state “wait for take off” and start a python script with: 
 
```
commander = scf.cf.high_level_commander
commander.takeoff(0.7,2.0)
time.sleep(30)
```
 
grab one of the hovering drone and bring it in in the near of the other drone
 
=>	Nothing happens :(
 
 
But if only one drone is hovering with the above mentioned python script and the other drone is executing the decentralized swarm demo and comes in the near of the script based hovering drone
 
=>	the python script based hovering drone is moving and avoiding a collision
 
 
I can´t figure out, why this collision avoidance isn´t working, when I use only the script based flying for both drones

**Comment by krichardsson** [comment, marked as answer]
I'm not completely sure but what I think is happening is that you are using both P2P and "normal" communication at the same time, the current protocol does not really support this use case. This is also why we use a stationary "sniffer" Crazyflie in the demo (connected through USB) that uses P2P to change settings in the swarm.

Each drone is using P2P to broadcast its position, the peers are using the position information to perform the collision avoidance. My guess is that when you run the script to connect to the Crazyflie, the P2P messages are blocked and the collision avoidance does not work. 

If I remember correctly, we broadcast the position all the time in the demo and the peers will get the position also when a drone is not flying, this would mean that you can simply carry one of them around and others would avoid it. There is one thing you have to change though; when the position is received from a peer it is passed on to `peerLocalizationTellPosition()` to let the collision avoidance system know where the peer is located. As can be seen in [ds_p2p_interface.c](https://github.com/bitcraze/crazyflie-firmware-experimental/blob/icra-2023/examples/demos/decentralized_swarm/src/common_files/ds_p2p_interface.c#L110 ) this is only done when the peer is in a flying state and you have to modify this to always call `peerLocalizationTellPosition()`, also when carried around. 

I think I tested this at some point and it was sort of fun to walk around in a swarm with a Crazyflie on my head and all other drones avoided me (the Crazyflie).

**Reply by heinrichwagner** [reply]
Thank you for your fast reply!

Is it possible to send flying commands from a python script (e.g. with high-level-commander) with the sniffer and the P2P protocol? 

Or can you imagine other solutions for controlling a swarm from a control station with the on-board collision avoidance enabled?

**Reply by krichardsson** [reply]
> Is it possible to send flying commands from a python script (e.g. with high-level-commander) with the sniffer and the P2P protocol?

Nothing is implemented for this.
You will have to modify the communication mechanism in ds_p2p_interface.h. In the struct `copter_message_t` you can find the `desiredFlyingCopters` member, this is currently the only "control data". You could add something here perhaps? You would have to define your own format though. 
The control data is designed to be propagated in the swarm by always adding the latest known version of control data to all outgoing messages.

> Or can you imagine other solutions for controlling a swarm from a control station with the on-board collision avoidance enabled?

Not really, I think the above mechanism is the way to go.

The problem (fun property) of this type of solution is that it is a bit tricky to debug. As long as one Crazyflie in the swarm is still alive, the data will survive and will be consumed by other Crazyflies that are booted. Unexpected things might happens and it can be hard to understand what state all members are in. Remember that you can connect to a copter using USB at the same time as you do P2P to examine logs and such, it is not a good solution while you are flying though ;-)

**Reply by heinrichwagner** [reply]
@krichardsson thank you for your help!

Adding the STATE_WAIT_FOR_TAKE_OFF to the if condition in [ds_p2p_interface](https://github.com/bitcraze/crazyflie-firmware-experimental/blob/c873bde6b9fa45118bc8f3d7bd736ce5f3b01e3e/examples/demos/decentralized_swarm/src/common_files/ds_p2p_interface.c#L92) worked for my use case (communication through Crazyradio).

I did just a quick test with two drones hovering with the script and both of them react if I grab one and bring it near the other. 

I will do a more detailed test with more drones flying at the same time and will report the results.
